### DISCUSSION TITLE
What's HistoryList actually for and can I use it for navigation?

### DISCUSSION THREAD

**Discussion by Ca1vino** [TiddlyWiki/TiddlyWiki5#5603] [Q&A]
I haven't found code examples for how to do this -- is it possible to make a button that triggers an action that pops the HistoryList stack and puts the focus back where it was last time?

Can one access the entries in the JSON array without writing javascript widgets or macros?

**Comment by pmario** [comment]
> Can one access the entries in the JSON array without writing javascript widgets or macros?

no.

**Comment by pmario** [comment]
>I haven't found code examples for how to do this -- is it possible to make a button that triggers an action that pops the HistoryList stack and puts the focus back where it was last time?

You can use HistoryList fields to visualize something, but navigation is done with tm-xxx messages. see: https://tiddlywiki.com/#Messages

**Reply by Ca1vino** [reply]
What I don't get is -- even if navigation is done with messages, how do you index and read the HistoryList to find out where to navigate TO?  Oddly, perhaps, that is where I got stuck, how to write code to read the damn thing.

I'd like to be able to reverse order on the recently visited tiddlers, like going backward on a bread crumbs trail.  

Or, perhaps just having a breadcrumb plugin would work, but I hope to keep the visual clutter to a minimum - too much of it intimidates some potential users.

**Reply by Jermolene** [reply]
> What I don't get is -- even if navigation is done with messages, how do you index and read the HistoryList to find out where to navigate TO? Oddly, perhaps, that is where I got stuck, how to write code to read the damn thing.

The $:/HistoryList tiddler is maintained by JS code, and the data is not in a format that is readily accessible to wikitext.

> I'd like to be able to reverse order on the recently visited tiddlers, like going backward on a bread crumbs trail.
> 
> Or, perhaps just having a breadcrumb plugin would work, but I hope to keep the visual clutter to a minimum - too much of it intimidates some potential users.

The approach taken by the core is to expose the history via the story list. For example, visit https://tiddlywiki.com/prerelease/, and then select the storyview "Stacked" in the sidebar "Tools" tab. Then navigate through a few tiddlers, and observe what happens when you close the frontmost one.

It should look like this:

<img width="1658" alt="image" src="https://user-images.githubusercontent.com/174761/115952132-49a14c00-a4dc-11eb-8bf5-82461c6623ab.png">


(To make the screenshot I also set "Display tiddler titles as links" in control panel, and adjusted the "fan separation" in the sidebar).

**Comment by twMat** [comment]
@Ca1vino - it is probably not what you mean, but I've added this to my *close* and *delete* buttons to navigate to the tiddler physically above the current.

`<$action-navigate $to={{{ [list[$:/StoryList!!list]before<currentTiddler>]}}}/>`

**Reply by Ca1vino** [reply]
Definitely a good trick, thanks.
